pageserver/src/tenant.rs

```diff
@@ -1189,6 +1189,19 @@ impl Tenant {
                 format!("Failed to load layermap for timeline {tenant_id}/{timeline_id}")
             })?;
 
+        // When unarchiving, we've mostly likely lost the heatmap generated prior
+        // to the archival operation. To allow warming this timeline up, generate
+        // a previous heatmap which contains all visible layers in the layer map.
+        // This previous heatmap will be used whenever a fresh heatmap is generated
+        // for the timeline.
+        if matches!(cause, LoadTimelineCause::Unoffload) {
+            if let Some(unarchival_heatmap) = timeline.generate_unarchival_heatmap().await {
+                timeline
```

Comment:
This also needs to recurse into ancestors: some archived timelines may have minimal data of their own, but depend on a lot of historical layers in their parent which have been evicted.

The ancestors' layer viisibility state will be stale at this point, so we should also be calling `update_layer_visibility` on the parent of the unarchived timeline.  We then add every visible layer from the parent into the parent's heatmap (this might over-estimate the set of layers we need), or we could try and calculate exactly which layers were made visible by this child timeline specifically.  Or split the difference and add "visible non-resident layers whose LSN is lower than the branch point" for a much simple calculation

Reply:
Done in https://github.com/neondatabase/neon/pull/10880/commits/d11809d813c02e7e850c6b7b81773280d3245817.

Went for: split the difference and add "visible non-resident layers whose LSN is lower than the branch point". However, we don't have to care if the layers are currently resident or not. If the layer in the previous heatmap
is actually resident we'll ignore it in the next call to `Timeline::generate_heatmap`.